### PR TITLE
add order by expression algorithm

### PR DESCRIPTION
**Pull request by Gustry**
## Description

Because the "easy method" here https://gis.stackexchange.com/a/218979 didn't convince me ;-)

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/analysis/processing/qgsalgorithmorderbyexpression.cpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+QVariantMap QgsOrderByExpressionAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  std::unique_ptr< QgsFeatureSource > source( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !source )
+    return QVariantMap();
+
+  QString expressionString = parameterAsExpression( parameters, QStringLiteral( "EXPRESSION" ), context );
+
+  bool ascending = parameterAsBool( parameters, QStringLiteral( "ASCENDING" ), context );
+  bool nullsFirst = parameterAsBool( parameters, QStringLiteral( "DROP_M_VALUES" ), context );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmorderbyexpression.cpp]: Fix parameter name

File: src/analysis/processing/qgsalgorithmorderbyexpression.cpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    return QVariantMap();
+
+  QgsFeatureRequest request;
+  request.addOrderBy( expressionString, ascending, nullsFirst );
+
+  QgsFeature inFeature;
+  QgsFeatureIterator features = source->getFeatures( request );
+  while ( features.nextFeature( inFeature ) )
+  {
+    sink->addFeature( inFeature );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmorderbyexpression.cpp]: Add feedback for progress add cancellation here... Someone might be sorting a million row table!

File: src/analysis/processing/qgsalgorithmorderbyexpression.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, sinkId, source->fields(), source->wkbType(), source->sourceCrs() ) );
+  if ( !sink )
+    return QVariantMap();
+
+  QgsFeatureRequest request;
+  request.addOrderBy( expressionString, ascending, nullsFirst );
+
+  QgsFeature inFeature;
+  QgsFeatureIterator features = source->getFeatures( request );
+  while ( features.nextFeature( inFeature ) )
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmorderbyexpression.cpp]: Fun fact (off topic) :

sink->addFeatures( features )

 also works

**Review by nyalldawson — COMMENTED** [review]
Looking good!

**Comment by nyalldawson** [comment]
@Gustry 

You need to cherry-pick  https://github.com/nyalldawson/QGIS/commit/f21de91f20e1eca8bbecef6064765054828369b2 to fix the test. Seems there's an extra unicode character this PR inserts into the test yaml file which breaks things.

**Comment by m-kuhn** [comment]
https://github.com/Gustry/QGIS/blob/b2a44983fbbb8ea136991cc3587b96a765baef33/python/plugins/processing/tests/README.md

**Comment by Gustry** [comment]
Thanks !

**Comment by Gustry** [comment]
The expected layer is ok in QGIS (input vs output ordered by intval)
screen shot 2017-11-28 at 09 07 16

If I run `ctest -V -R ProcessingQgisAlgorithmsTest` it seems fine for this algorithm. ( I got a lot of other algorithms which fail on mac).

But travis is not happy:
```
======================================================================
[33mFAIL: test_algorithms:212
[0m[33m'Order by expression', {'algorithm': 'native:orderbyexpression', 'name': 'Order (__main__.TestQgisAlgorithms)
[0m[33m----------------------------------------------------------------------
[0m[33mTraceback (most recent call last):
[0m[33m  File "/root/QGIS/python/plugins/processing/tests/AlgorithmsTestBase.py", line 135, in check_algorithm
[0m[33m    self.check_results(results, context, defs['params'], defs['results'])
[0m[33m  File "/root/QGIS/python/plugins/processing/tests/AlgorithmsTestBase.py", line 292, in check_results
[0m[33m    self.assertLayersEqual(expected_lyrs[0], result_lyr, compare=compare, pk=pk)
[0m[33m  File "/root/QGIS/build/output/python/qgis/testing/__init__.py", line 58, in assertLayersEqual
[0m[33m    self.checkLayersEqual(layer_expected, layer_result, True, **kwargs)
[0m[33m  File "/root/QGIS/build/output/python/qgis/testing/__init__.py", line 191, in checkLayersEqual
[0m[33m    repr(attr_result)
[0m[33mAssertionError: '33' != 33 : Features 0/0 differ in attributes
[0m[33m
[0m[33mRunning alg: "qgis:distancematrix"
[0m[33mRunning alg: "qgis:distancematrix"
[0m[33mRunning alg: "qgis:union"
[0m[33m
[0m[33m * Field expected: intval (String)
[0m[33m * result  : intval (String)
[0m[33m
[0m[33m * Expected: '33' != Result  : 33
[0m[33m
[0m[33m----------------------------------------------------------------------
[0mRan 278 tests in 14.031s
```

File: src/analysis/processing/qgsalgorithmorderbyexpression.cpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  addParameter( new QgsProcessingParameterExpression( QStringLiteral( "EXPRESSION" ), QObject::tr( "Expression" ), QVariant(), QStringLiteral( "INPUT" ) ) );
+  addParameter( new QgsProcessingParameterBoolean( QStringLiteral( "ASCENDING" ), QObject::tr( "Ascending" ), true ) );
+  addParameter( new QgsProcessingParameterBoolean( QStringLiteral( "NULLS_FIRST" ), QObject::tr( "Nulls first" ), false ) );
+
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "OUTPUT" ), QObject::tr( "Output layer" ) ) );
+}
+
+QString QgsOrderByExpressionAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm sorts a vector layer according to an expression." );
```
Comment (m-kuhn) [inline src/analysis/processing/qgsalgorithmorderbyexpression.cpp]: I think here should also a note be added, that it's up to each provider if this order is preserved and that relying on this is often exploiting undocumented behavior. (in fact, I think this algorithm will only unleash its full potential within the specs once we have feature pipelines in place).

Comment (Gustry) [reply src/analysis/processing/qgsalgorithmorderbyexpression.cpp]: Thanks @m-kuhn I saw your comment about ordering yesterday https://gis.stackexchange.com/questions/263130/render-polygons-ordered-by-size/263133#263133

I tried on some files (geojson) and it seems to work, but you are saying it's not expected for now? Even with postgres provider? Do you have an idea of which providers might respect the order which providers don't?

Comment (m-kuhn) [reply src/analysis/processing/qgsalgorithmorderbyexpression.cpp]: I think many will do it to some degree, especially simple transfer files like geojson, but I think it's exploiting undocumented behavior and may change depending on context, indexing, ...

**Comment by nyalldawson** [comment]
Merging as per the result of https://www.loomio.org/d/38Aiya0q/3-0-soft-freeze-exemptions

**Pull request merged**